core/src/main/java/kafka/server/handlers/DescribeTopicPartitionsRequestHandler.java

```diff
@@ -0,0 +1,142 @@
+    MetadataCache metadataCache;
+    AuthHelper authHelper;
+    KafkaConfig config;
+
+    public DescribeTopicPartitionsRequestHandler(
+        MetadataCache metadataCache,
+        AuthHelper authHelper,
+        KafkaConfig config
+    ) {
+        this.metadataCache = metadataCache;
+        this.authHelper = authHelper;
+        this.config = config;
+    }
+
+    public DescribeTopicPartitionsResponseData handleDescribeTopicPartitionsRequest(RequestChannel.Request abstractRequest) {
+        if (metadataCache instanceof ZkMetadataCache) {
+            throw new InvalidRequestException("ZK cluster does not handle DescribeTopicPartitions request");
+        }
+        KRaftMetadataCache kRaftMetadataCache = (KRaftMetadataCache) metadataCache;
+
+        DescribeTopicPartitionsRequestData request = ((DescribeTopicPartitionsRequest) abstractRequest.loggableRequest()).data();
+        Set<String> topics = new HashSet<>();
+        boolean fetchAllTopics = request.topics().isEmpty();
+        DescribeTopicPartitionsRequestData.Cursor cursor = request.cursor();
+        String cursorTopicName = cursor != null ? cursor.topicName() : "";
+        if (fetchAllTopics) {
+            JavaConverters.asJava(kRaftMetadataCache.getAllTopics()).forEach(topicName -> {
+                if (topicName.compareTo(cursorTopicName) >= 0) {
+                    topics.add(topicName);
+                }
+            });
+        } else {
+            request.topics().forEach(topic -> {
+                String topicName = topic.name();
+                if (topicName.compareTo(cursorTopicName) >= 0) {
+                    topics.add(topicName);
+                }
+            });
+
+            if (cursor != null && !topics.contains(cursor.topicName())) {
+                // The topic in cursor must be included in the topic list if provided.
+                throw new InvalidRequestException("DescribeTopicPartitionsRequest topic list should contain the cursor topic: " + cursor.topicName());
+            }
+        }
+
+        // Do not disclose the existence of topics unauthorized for Describe, so we've not even checked if they exist or not
+        Set<DescribeTopicPartitionsResponseTopic> unauthorizedForDescribeTopicMetadata = new HashSet<>();
+
+        Stream<String> authorizedTopicsStream = topics.stream().filter(topicName -> {
+            boolean isAuthorized = authHelper.authorize(
+                abstractRequest.context(), DESCRIBE, TOPIC, topicName, true, true, 1);
+            if (!fetchAllTopics && !isAuthorized) {
+                // We should not return topicId when on unauthorized error, so we return zero uuid.
+                unauthorizedForDescribeTopicMetadata.add(describeTopicPartitionsResponseTopic(
+                    Errors.TOPIC_AUTHORIZATION_FAILED, topicName, Uuid.ZERO_UUID, false, Collections.emptyList())
+                );
+            }
+            return isAuthorized;
+        }).sorted();
```

Comment:
This `sorted` will traverse the whole stream, meaning we'll call the authorizer for each topic. Let's move the sorting before the filter so the downstream consumer of this stream (KRaftMetadataCache#getTopicMetadataForDescribeTopicResponse) will only need to hit the authorize enough times to generate the required number of partitions.

Reply:
Thanks for the tip!